pkg/services/ngalert/state/manager.go

```diff
@@ -150,18 +149,69 @@ func (st *Manager) Get(orgID int64, alertRuleUID, stateId string) *State {
 	return st.cache.get(orgID, alertRuleUID, stateId)
 }
 
-// ResetStateByRuleUID deletes all entries in the state manager that match the given rule UID.
-func (st *Manager) ResetStateByRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey) []*State {
-	logger := st.log.New(ruleKey.LogContext()...)
-	logger.Debug("Resetting state of the rule")
-	states := st.cache.removeByRuleUID(ruleKey.OrgID, ruleKey.UID)
+// ResetStateByDeletedRuleUID deletes all entries in the state manager that match the given rule UID.
+func (st *Manager) ResetStateByDeletedRuleUID(ctx context.Context, rule *ngModels.AlertRule) []*State {
```

Comment:
I see there is a lot of duplication between the two functions. This is what I was thinking, what do you think?

```
func (st *Manager) ResetStateByRuleUID(ctx context.Context, rule *ngModels.AlertRule, resetReason string, appendStateTs []StateTransition) []*State {
	key := rule.GetKey()
	logger := st.log.New(key.LogContext()...)
	logger.Debug("Resetting rule state", "reason", resetReason)

	states := st.cache.removeByRuleUID(rule.OrgID, rule.UID)
	if len(states) > 0 && st.instanceStore != nil {
		err := st.instanceStore.DeleteAlertInstancesByRule(ctx, key)
		if err != nil {
			logger.Error("Failed to delete states that belong to a deleted rule from database", "error", err)
		}
	}

	transitions := make([]StateTransition, 0, len(states))
	for _, s := range states {
		transitions = append(transitions, StateTransition{
			State:               nil,
			PreviousState:       s.State,
			PreviousStateReason: s.StateReason,
		})
		transitions = append(transitions, appendStateTs...)
	}

	if st.historian != nil {
		st.historian.RecordStatesAsync(ctx, rule, transitions)
	}

	logger.Info("Rule state was reset", "reason", resetReason, "states", len(states))
	return states
}
```

Reply:
Agree with George, let's use one method.